scripts/get_orphaned_functions.py

```diff
@@ -0,0 +1,243 @@
+    """
+    Check if the function from a specific module is utilized in any other module.
+
+    Args:
+    function_name (str): The name of the function to check.
+    module_name (str): The name of the module where the function is originally defined.
+
+    Returns:
+    bool: True if the function is utilized externally, False otherwise.
+    """
+    function_pattern = re.compile(r"\b" + re.escape(function_name) + r"\b")
+    func_location_module = os.path.join(main_dir, "modules", module_name)
+    search_directories = [
+        os.path.join(main_dir, "modules"),
+        os.path.join(main_dir, "models"),
+        os.path.join(main_dir, "scripts"),
+        os.path.join(main_dir, "base"),
+        os.path.join(main_dir, "tests"),
+        os.path.join(main_dir, "base"),
+    ]
+
+    # print(color_codes[2] + "Function name:",function_name," \t Function's Location modules :",func_location_module,ENDC,)
+
+    for directory in search_directories:
+        for root, dirs, files in os.walk(directory):
+            if (
+                root.__contains__(func_location_module)
+                or root.__contains__("/man")
+                or root.__contains__("/vignettes")
+            ):
+                # print(f"{color_codes[13]} Skipping the module directory {root}{ENDC}")
+                continue
+            else:
+                # print(f"{color_codes[14]} Searching for the function in the directory {root}{ENDC}")
+                for file in files:
+                    if file.endswith(".R"):
+                        file_path = os.path.join(root, file)
+                        with open(file_path, "r") as file:
+                            file_content = file.read()
+                            if function_pattern.search(file_content):
+                                # print(f'{color_codes[12]} Function "{function_name}" is utilized EXTERNALLY in file: {file_path}{ENDC}')
+                                return True
+    return False
+
+
+def print_in_batches(list_to_print, label):
+    # print(label)
+    color_index = 0  # Start with the first color
+    for i in range(0, len(list_to_print), 5):
+        batch = list_to_print[i : i + 5]
+        # Get the current color
+        color = color_codes[color_index % len(color_codes)]
+        # Print the batch in the current color
+        print(color + ", ".join(map(str, batch)) + ENDC)
+        # Move to the next color for the next batch
+        color_index += 1
+
+
+def add_to_csv(list_1, list_2, csv_file_path):
+    with open(csv_file_path, mode="w", newline="") as file:
+        writer = csv.writer(file)
+
+        # Writing the header for list_1
+        writer.writerow(["Purely Orphaned functions", "module", "function_name"])
+
+        # Writing data from list_1
+        for module, function in list_1:
+            writer.writerow(["", module, function])
+
+        # Writing the header for list_2
+        writer.writerow(["Externally Utilized Orphaned Functions"])
```

Comment:
I'm puzzled by this classification -- if a function is used externally, in what sense is it orphaned?

Reply:
The Function would be orphaned within it's `OWN` directory while at the same time it is being utilised externally. These are functions that are not used within their own module (hence considered `orphaned` in a local sense) but are called or used by external `packages` that depend on the prior module. These functions might not be directly referenced within the project's codebase but are still necessary for the functionality provided to external consumers. I've some of my own doubts regarding this. Many a times, we'll see such functions' utilization in external `packages` as a coupled dependency within two (or more) different packages. Shifting functions that are identified as "externally utilized orphaned functions" to external packages or creating a new directory to handle all coupled dependencies within different packages in the repository can be a good approach.

Reply:
There's no requirement that a function be called within its own package to be considered useful, so I strongly suggest not referring to these as "orphaned".  I take your point that "defined only here, used only there" might be a candidate for relocation, but my hunch is these will be a small fraction of the functions identified by this criterion.

But! A function that's not called within its own package is very likely _untested_, so it may be worth identifying them on those grounds. But if we want to go into that level of detail it may be worth moving from regexes to a parsing approach that can give more information / allow more control over the exact types of call matched.

Reply:
I think it'd just warn against trying to over-engineer a script to try and solve multiple different use cases and problems, and instead just focus on the question being asked (are there functions that aren't getting used). Personally, I wouldn't rely on just this function alone to make decisions about whether to delete, deprecate, or move a function, but would instead follow up on proposed orphans (e.g. using grep) to see how and where it is being used.

Per @infotroph point about tests, that cuts both ways -- there could be functions that were designed with lots of tests, but are never used OUTSIDE of their build tests, that would also be considered orphans.